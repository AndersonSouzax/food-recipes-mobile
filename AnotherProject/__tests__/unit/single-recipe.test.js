import 'react-native';
import React from 'react';
import { render, waitFor, act, fireEvent, 
	waitForElementToBeRemoved } from '@testing-library/react-native';

import SingleRecipe from '../../src/single-recipe';

import { Provider as PaperProvider } from 'react-native-paper';

const mockRoute = { 
	params: { 
		recipe : { id: 1, title : 'Fish', user : { id: 1 },
      description : 'you got that yummi yummi...',
      category : { id : 1, name : "Japonese Food", 
          image : 'https://upload.wikimedia.org/wikipedia/commons/5/57/Oseti.jpg' 
      } 
    },
    user : { id : 1 }
  }
};

jest.setTimeout(20000);

jest.mock('react-native/Libraries/Animated/src/NativeAnimatedHelper');

jest.mock('../../src/services/api');

test('Renders correctly', () => {
  const component = render(
  	<PaperProvider>
  		<SingleRecipe route={mockRoute}/>
  	</PaperProvider>
  );
  const componentJS = component.toJSON();
  expect(componentJS).toMatchSnapshot();
});

test('Go back to the recipes list', () => {

	const navigationMock = { goBack: jest.fn() };

	const { getByTestId } = render(
		<PaperProvider>
			<SingleRecipe navigation={navigationMock} route={mockRoute} />
		</PaperProvider>
	);

	const backButton = getByTestId('backButton');

	fireEvent.press(backButton);

	expect(navigationMock.goBack.mock.calls.length).toBe(1);

});

test('Make user\'s recipe editable', async () => {

	const component = render(
		<PaperProvider>
			<SingleRecipe route={mockRoute} />
		</PaperProvider>
	);

	const { getByTestId } = component;

	const options = getByTestId('options');

	fireEvent.press(options);

	await act(async () => {	

		await waitFor(() => getByTestId('editReciple'));
		
		const editButton = getByTestId('editReciple');

		fireEvent.press(editButton);

		await act(async () => {

			/*
				* The categories must be loaded only when edit mode is selected
				*/
			await waitFor(() => getByTestId('categoryList'));
			
			const desc = getByTestId('description');

			/*
				* Test the informations changing...
				*/

			const event = { target : { testID : 'description', value : 'loveyou' } };

			fireEvent.changeText(desc, event);
			
			await waitFor(() => getByTestId('description').props.value === 'loveyou');

		});

	});

});

test('Deleting recipe correctly', async () => {

	const navigationMock = { goBack: jest.fn() };

	const component = render(
		<PaperProvider>
			<SingleRecipe route={mockRoute} navigation={navigationMock}/>
		</PaperProvider>
	);

	const { getByTestId } = component;

	const options = getByTestId('options');

	fireEvent.press(options);

	await act(async () => {	

		await waitFor(() => getByTestId('deleteReciple'));
		
		const deleteButton = getByTestId('deleteReciple');

		fireEvent.press(deleteButton);

		await act(async () => {

			//await waitFor(() => getByTestId('confirmDelete'));

			const confirm = getByTestId('confirmDelete');

			await fireEvent.press(confirm);

			//await act(async () => {

 				//await waitFor(() => getByTestId('loadElement'));

 				// await waitForElementToBeRemoved(() => getByTestId('loadElement'));

				// The test gets too weak if only this call is checked...
				expect(navigationMock.goBack.mock.calls.length).toBe(1);

			//});
		});
	});
});

test('Updating recipe correctly', async () => {

	const component = render(
		<PaperProvider>
			<SingleRecipe route={mockRoute} />
		</PaperProvider>
	);

	const { getByTestId } = component;

	const options = getByTestId('options');

	const saveButton = getByTestId('saveButton');

	fireEvent.press(options);

	await waitFor(() => getByTestId('editReciple'));
	
	const editButton = getByTestId('editReciple');

	fireEvent.press(editButton);

	await waitFor(() => getByTestId('description'));

	const desc = getByTestId('description');

	const event = { target : { testID : 'description', value : 'loveyou' } };

	fireEvent.changeText(desc, event);
	
	await waitFor(() => getByTestId('description').props.value === 'loveyou');
	
	await fireEvent.press(saveButton);

	/* The value of view description must be the same that was inserted during editing */
	await waitFor(() => getByTestId('descView').props.value === 'loveyou');

});



