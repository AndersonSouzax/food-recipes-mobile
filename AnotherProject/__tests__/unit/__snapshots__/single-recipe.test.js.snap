// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Array [
        Object {
          "backgroundColor": "#EA1D2C",
          "elevation": 4,
          "zIndex": 0,
        },
        Object {
          "shadowColor": "#000000",
          "shadowOffset": Object {
            "height": 3,
            "width": 0,
          },
          "shadowOpacity": 0.24,
          "shadowRadius": 4,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#EA1D2C",
          "elevation": 0,
          "flexDirection": "row",
          "height": 56,
          "justifyContent": "space-between",
          "marginTop": 0,
          "paddingHorizontal": 4,
        }
      }
    >
      <View
        accessibilityLabel="Back"
        accessibilityRole="button"
        accessibilityState={
          Object {
            "disabled": undefined,
          }
        }
        accessible={true}
        focusable={true}
        hitSlop={
          Object {
            "bottom": 6,
            "left": 6,
            "right": 6,
            "top": 6,
          }
        }
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "overflow": "hidden",
            },
            Array [
              Object {
                "alignItems": "center",
                "justifyContent": "center",
                "margin": 6,
                "overflow": "hidden",
              },
              Object {
                "borderRadius": 18,
                "height": 36,
                "width": 36,
              },
              undefined,
              undefined,
            ],
          ]
        }
        testID="backButton"
      >
        <View>
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "justifyContent": "center",
                },
                Object {
                  "height": 24,
                  "width": 24,
                },
              ]
            }
          >
            <View
              style={
                Object {
                  "bottom": 0,
                  "left": 0,
                  "opacity": 1,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                  "transform": Array [
                    Object {
                      "rotate": "0deg",
                    },
                  ],
                }
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "justifyContent": "center",
                    },
                    Object {
                      "height": 24,
                      "transform": Array [
                        Object {
                          "scaleX": 1,
                        },
                      ],
                      "width": 24,
                    },
                  ]
                }
              >
                <Image
                  source={
                    Object {
                      "process": [Function],
                    }
                  }
                  style={
                    Array [
                      Object {
                        "height": 21,
                        "resizeMode": "contain",
                        "width": 21,
                      },
                      Object {
                        "tintColor": "#ffffff",
                      },
                    ]
                  }
                />
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        collapsable={false}
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "disabled": false,
            }
          }
          accessible={true}
          focusable={true}
          hitSlop={
            Object {
              "bottom": 6,
              "left": 6,
              "right": 6,
              "top": 6,
            }
          }
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Array [
                Object {
                  "alignItems": "center",
                  "justifyContent": "center",
                  "margin": 6,
                  "overflow": "hidden",
                },
                Object {
                  "borderRadius": 18,
                  "height": 36,
                  "width": 36,
                },
                false,
                Object {
                  "color": "white",
                },
              ],
            ]
          }
          testID="options"
        >
          <View>
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "justifyContent": "center",
                  },
                  Object {
                    "height": 24,
                    "width": 24,
                  },
                ]
              }
            >
              <View
                style={
                  Object {
                    "bottom": 0,
                    "left": 0,
                    "opacity": 1,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                    "transform": Array [
                      Object {
                        "rotate": "0deg",
                      },
                    ],
                  }
                }
              >
                <Text
                  accessibilityElementsHidden={true}
                  allowFontScaling={false}
                  importantForAccessibility="no-hide-descendants"
                  pointerEvents="none"
                  style={
                    Array [
                      Object {
                        "color": "rgba(0, 0, 0, 0.54)",
                        "fontSize": 24,
                      },
                      Array [
                        Object {
                          "transform": Array [
                            Object {
                              "scaleX": 1,
                            },
                          ],
                        },
                        Object {
                          "backgroundColor": "transparent",
                        },
                      ],
                      Object {
                        "fontFamily": "Material Design Icons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      Object {},
                    ]
                  }
                >
                  ó°‡™
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTSafeAreaView>
  <View
    style={
      Object {
        "alignItems": "center",
      }
    }
  >
    <Text
      selectable={true}
    >
      Fish
    </Text>
    <Text>
      Category: 
      Japonese Food
    </Text>
    <View
      style={
        Object {
          "backgroundColor": "#ffffff",
          "borderRadius": 4,
          "elevation": 1,
          "shadowColor": "#000000",
          "shadowOffset": Object {
            "height": 0.5,
            "width": 0,
          },
          "shadowOpacity": 0.24,
          "shadowRadius": 0.75,
          "width": "80%",
        }
      }
    >
      <View
        accessible={true}
        focusable={false}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "flexGrow": 1,
            "flexShrink": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "backgroundColor": "#eeeeee",
                "height": 195,
                "overflow": "hidden",
              },
              Object {
                "borderRadius": 4,
              },
              undefined,
            ]
          }
        >
          <Image
            siblings={
              Array [
                "withTheme(Card.Cover)",
              ]
            }
            source={
              Object {
                "uri": "https://upload.wikimedia.org/wikipedia/commons/5/57/Oseti.jpg",
              }
            }
            style={
              Array [
                Object {
                  "flex": 1,
                  "height": undefined,
                  "justifyContent": "flex-end",
                  "padding": 16,
                  "resizeMode": "cover",
                  "width": undefined,
                },
                Object {
                  "borderRadius": 4,
                },
              ]
            }
          />
        </View>
      </View>
    </View>
    <Text
      selectable={true}
      testID="descView"
    >
      you got that yummi yummi...
    </Text>
  </View>
</View>
`;
